torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -460,44 +452,229 @@ void CompiledFusionFunction::launch_with_tensors(at::ArrayRef<at::Tensor> inputs
+    concat_desc = codegen::emitCompilationUnit(cu, name, agraph, true);
+    compilation_unit = cu.str();
+    nvrtcProgram program;
+    TORCH_NVRTC_CHECK(nvrtcCreateProgram(&program, compilation_unit.c_str(), NULL, 0, nullptr, nullptr));
+
+    std::string compute = "--gpu-architecture=compute_" + std::to_string(prop.major) + std::to_string(prop.minor);
+    std::vector<const char *> args = {"--std=c++11", compute.c_str()};
+    nvrtcResult result = nvrtcCompileProgram(program, args.size(), args.data());
+    if (result == NVRTC_ERROR_COMPILATION) {
+      size_t logsize;
+      nvrtcGetProgramLogSize(program, &logsize);
+      std::vector<char> log(logsize);
+      nvrtcGetProgramLog(program, log.data());
+      cu << log.data();
+      throw std::runtime_error(cu.str());
+    }
+    ResourceGuard holdProgram([&] {
+      TORCH_NVRTC_CHECK(nvrtcDestroyProgram(&program));
+    });
+    TORCH_NVRTC_CHECK(result);
+
+    size_t ptx_size;
+    TORCH_NVRTC_CHECK(nvrtcGetPTXSize(program, &ptx_size));
+    ptx.resize(ptx_size);
+    TORCH_NVRTC_CHECK(nvrtcGetPTX(program, ptx.data()));
+
+    TORCH_CU_CHECK(cuModuleLoadData(&module, ptx.data()));
+    TORCH_CU_CHECK(cuModuleGetFunction(&function, module, name.c_str()));
+
+    TORCH_CU_CHECK(cuOccupancyMaxActiveBlocksPerMultiprocessor(
+      &maxBlocks, function, 128, 0));
+    maxBlocks *= prop.multiProcessorCount;
+  }
+  virtual ~CUDAFusionFunction() override {
+    TORCH_CU_CHECK(cuModuleUnload(module));
+  }
+protected:
+  virtual at::Backend backend() const override {
+    return at::kCUDA;
+  }
+  virtual void launch_raw(uint32_t numel, void ** arguments) override {
+     int numBlocks = std::min(maxBlocks, ceilDiv(numel, blockSize));
+     //std::cout << "maxBlocks = " << maxBlocks << " needed blocks: " << ceilDiv(numel,blockSize)
+     //          << " numblocks =  " << numBlocks;
+
+     // it is possible that this is the first cuda call on this thread
+     // so make sure we initialize the Driver API's context
+     // cudaFree(0) accomplishes this.
+     cudaFree(0);
+
+     TORCH_CU_CHECK(cuLaunchKernel(
+       function,
+       numBlocks, 1, 1,
+       blockSize, 1, 1,
+       0, nullptr,
+       arguments,
+       nullptr));
+  }
+  std::vector<char> ptx;
+  CUmodule module;
+  CUfunction function;
+
+  // we record prop/device so if they are availiable for launch heuristics
+  // querying at launch is too slow for device properties.
+  int device;
+  cudaDeviceProp prop;
+  int blockSize = 128;
+  int maxBlocks;
+};
+
+#endif
+
+struct TempFile {
+  TH_DISALLOW_COPY_AND_ASSIGN(TempFile);
+  TempFile(const std::string & t, int suffix) {
+    std::vector<char> tt(t.c_str(), t.c_str() + t.size() + 1);
+    int fd = mkstemps(tt.data(), suffix);
+    JIT_ASSERT(fd != -1);
+    file_ = fdopen(fd, "r+");
+    name_ = std::string(tt.begin(), tt.end() - 1);
+  }
+  const std::string & name() const {
+    return name_;
+  }
+  void sync() {
+    fflush(file_);
+    fsync(fileno(file_));
```

Comment:
I don't see why you need to force the file to be durably flushed to the physical disk, since it's a temporary file anyway. All you need is for another process to be able to see it and for that fflush is sufficient.

Reply:
so `fflush` flushes FILE to to the file descriptor. Does a `write` ensure another processes read will succeed -- can you point me to that doc?